### ISSUE TITLE
WSL2: init process consumes 100%

### ISSUE THREAD

**Issue by firewave** [microsoft/WSL#5285]
I updated by Windows installation to 2004 and converted my existing Kali installation to WSL 2. Just leaving it running causes an `init` process to consume 100% CPU all the time. I already closed all applications which make use of my WSL instance (WSL shells, Docker and CLion) and the issue persists.

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
15408 root      20   0    1580    564     16 R  98.0   0.0 150:56.80 init
```

When attaching `strace` to that process it just generates an empty log:
```
strace: Process 15408 attached
strace: [ Process PID=15408 runs in x32 mode. ]
```

My Windows version is 10.0.19041.264.

**Comment by onomatopellan** [comment]
Try `wsl.exe --shutdown` to really shutdown the WSL2 VM.

Also if possible I recommend to uninstall the old Kali distro, run `wsl.exe  --set-default-version 2` and install again Kali from the store.

**Comment by firewave** [comment]
I would like to find out what the `init` process is actually doing. As I mentioned strace gave me nothing and I have no idea what else to do.

I had these issues in the past, but I couldn't keep that state permanent. I still have the same process sitting there after seven days now and could still get any additional information.

**Comment by benhillis** [comment]
@firewave - An empty strace usually means the program is spinning in usermode, typically a loop with a bad exit condition.  Do you have any more detailed instructions so I can try to reproduce this locally?  Any output in dmesg?

Thanks!

**Comment by firewave** [comment]
Thanks for the information.

No idea how to reproduce it - it suddenly was like this. In the future I might come across something, but because of several issues I have to revert to WSL 1 soon, so I wanted to get all the possible information as long as I can.

```
...
[   25.144448] init: (1) ERROR: UpdateTimezone:97: Europe/Berlin timezone not found. Is the tzdata package installed?
[   25.144455] init: (1) ERROR: InitEntryUtilityVm:2425: UpdateTimezone failed
...
[   25.562415] init: (2) ERROR: UtilCreateProcessAndWait:474: /bin/mount failed with 2
[   25.562560] init: (1) ERROR: UtilCreateProcessAndWait:489: /bin/mount failed with status 0x
[   25.562563] ff00
[   25.562570] init: (1) ERROR: ConfigMountFsTab:2077: Processing fstab with mount -a failed.
[   25.563644] init: (3) ERROR: UtilCreateProcessAndWait:474: /bin/mount failed with 2
[   25.563838] init: (1) ERROR: UtilCreateProcessAndWait:489: /bin/mount failed with status 0x
[   25.563841] ff00
[   25.563848] init: (1) ERROR: MountPlan9:478: mount cache=mmap,noatime,trans=fd,rfdno=8,wfdno=8,msize=65536,aname=drvfs;path=C:\;uid=0;gid=0;symlinkroot=/mnt/
[   25.563851]  failed 17
[   25.564774] init: (4) ERROR: UtilCreateProcessAndWait:474: /bin/mount failed with 2
...
[   42.618780] init: (118) ERROR: LogUsage:253: Could not find owner of socket 2 2222 18438
...
[  491.975531] init: (118) ERROR: LogUsage:253: Could not find owner of socket 2 2222 18438
...
[ 1131.798822] init: (118) ERROR: LogUsage:253: Could not find owner of socket 2 2222 18438
...
[14849.668061] print_req_error: I/O error, dev sdd, sector 0
...
[590035.586381] init: (118) ERROR: LogUsage:253: Could not find owner of socket 2 2222 18438
[590262.296336] init: (118) ERROR: LogUsage:253: Could not find owner of socket 2 2222 18438
[592705.770788] traps: init[12223] general protection ip:22b20a sp:7ffdf0a71590 error:0 in init[225000+6f000]
```

I assume the socket ones are related to `sshd` being restarted when I run `apt` updates. The ones at 25.x seem to early though the timezone one looks interesting. So it seems if anything in the log might be related to this issue it might be the I/O error.

FYI the system is working "fine" aside from the one `init` process consuming all the CPU.

**Comment by onomatopellan** [comment]
To rule out a faulty user profile try `wsl.exe --shutdown` and launch Kali with `wsl.exe -d Kali -u root`. Does it still happens?

**Comment by tuhlmann** [comment]
I experience the same thing on my wsl2 Ubuntu 18.04 distro.
It _seems_ to be related to a process from Windows, like IntelliJ Idea, to scan through its project files and re-index them, at least these are the occasions at which I observed this behavior. Shutting down Idea doesn't help. 

Also closing every terminal window or user process inside wsl2 doesn't help. `wsl --shutdown` and then restarting it does help, and the problem would not appear right away, just every so often. 

Sorry that I can't be more specific...

**Comment by alf-ytakada** [comment]
Hi.
I had the same problem after updating to 2004.
On my environment, root cause was Docker for Desktop.
When I uninstall/reinstall Docker for Desktop, I resolved.

**Comment by sjoller** [comment]
I too have issues with rogue /init processes consuming ~100% CPU, and I'm running an IntelliJ product as well.

My "solution", so far, has been to kill the runaway processes using htop, so a little faster than restarting WSL, but not really a useful solution in the long run, as the offending processes seem to reappear.

**Comment by jamestut** [comment]
Also happening here on build 21390.2025 and 22000.100 (both on WSL2 `Kernel version: 5.10.43`). Happens with both WSLg enabled and disabled, and both with and without Docker desktop. Workload is using two VSCode debugging sessions and several Windows Terminal. Only access the 9P (`\\wsl.localhost`) occasionally when doing Git. When it happens, it just happens randomly and without any adverse effect (besides the rogue `init` process maxing out a single core CPU time). In this particular instance, it happens well after 3 hours of WSL2 uptime (and keep going strong for around 70 minutes with no sign of stopping!).

Also interestingly, the rogue `init` process has a PID other than 1! So I can kill it easily using the standard `kill -9` command as root. Still annoying though esp. if using the laptop on battery.

Here is the output of `ps` when it happened.

```
# > top -d 1
top - 04:23:14 up 12:19,  0 users,  load average: 1.11, 1.27, 1.51
Tasks:  55 total,   2 running,  53 sleeping,   0 stopped,   0 zombie
%Cpu(s): 12.9 us,  0.1 sy,  0.0 ni, 86.8 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
MiB Mem :  15932.5 total,  10630.0 free,   4438.1 used,    864.4 buff/cache
MiB Swap:   4096.0 total,   4096.0 free,      0.0 used.  11225.2 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
26502 root      20   0    2468    532      0 R 100.0   0.0  72:53.41 init
17226 user      20   0 1050284 136528  19384 S   2.0   0.8   5:15.55 python3
21651 user      20   0  697468  56224  34836 S   1.0   0.3   0:31.09 node
    1 root      20   0    2116   1468   1016 S   0.0   0.0   3:42.74 init
    7 root      20   0    1752     80      0 S   0.0   0.0   0:00.00 init
    8 root      20   0    1752     88      0 S   0.0   0.0   0:00.43 init
    9 user      20   0  244160   7988   6244 S   0.0   0.0   0:00.70 fish
   79 root      20   0    2364    480      0 S   0.0   0.0   0:00.00 init
   80 root      20   0    2364    480      0 S   0.0   0.0   0:00.19 init
   81 user      20   0   87744   6476   5692 S   0.0   0.0   0:00.00 fish
   83 user      20   0    2624    768    680 S   0.0   0.0   0:00.00 sh
   84 user      20   0    2624    776    688 S   0.0   0.0   0:00.00 sh
   90 user      20   0    2624    776    688 S   0.0   0.0   0:00.00 sh
   92 user      20   0  931860  75384  32020 S   0.0   0.5   1:02.07 node
  103 user      20   0  641720  41440  29848 S   0.0   0.3   1:22.61 node
  114 user      20   0  880532  40780  29556 S   0.0   0.2   0:01.65 node
  134 user      20   0 1026732 165740  39236 S   0.0   1.0   2:53.24 node
  146 user      20   0  178420   7268   5948 S   0.0   0.0   0:00.07 fish
  227 user      20   0 1004504 271784  29836 S   0.0   1.7   0:46.21 node
  253 user      20   0   12.2g   1.5g  39244 S   0.0   9.5  27:30.13 node
  264 user      20   0  925232  64252  29784 S   0.0   0.4   0:10.50 node
```

As an additional note, it only happens on a single distro (in this case it is Ubuntu 21.04 server WSL image). I got several distros here, all started, but not actively used: only the Ubuntu 21.04 was used actively. Other distros are chilling at 0% CPU usage.

**Comment by onomatopellan** [comment]
@jamestut With `htop` in tree mode you can see which process was attached to that init, probably a `node` one.

**Comment by sorgloomer** [comment]
I have this issue about once a day. I also use JetBrains products, and the project is on the WSL filesystem, opened via `\\wsl$\Ubuntu-20.04\home\person\workspace\...` path. I usually just let the instance be started automatically by filesystem accesses.

`htop` only shows stacked init processes.

![image](https://user-images.githubusercontent.com/5985835/127974371-deda1a75-869c-4802-9fd4-0954f88285ba.png)

I do not have a definitive repro yet, but it happens quite often to me. Let me know I anybody here has some idea what we could check.

**Comment by sorgloomer** [comment]
I attached GDB to one of the runaway init processes, it looks like its stuck in a very short loop, around:

![image](https://user-images.githubusercontent.com/5985835/129176810-8b277c27-19d5-4b45-8e90-56cd514348ae.png)

```
(gdb) c
Continuing.
^C
Program received signal SIGINT, Interrupt.
0x000000000022ae8f in ?? ()
(gdb) bt
#0  0x000000000022ae8f in ?? ()
#1  0x000000000125fd30 in ?? ()
#2  0x00000000000000a0 in ?? ()
#3  0x0000000000000004 in ?? ()
#4  0x0000000000000090 in ?? ()
#5  0x0000000001259390 in ?? ()
#6  0x0000000001259410 in ?? ()
#7  0x00000000012593a0 in ?? ()
#8  0x0000000000000080 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb) c
Continuing.
^C
Program received signal SIGINT, Interrupt.
0x000000000022ae20 in ?? ()
(gdb) bt
#0  0x000000000022ae20 in ?? ()
#1  0x000000000125fd30 in ?? ()
#2  0x00000000000000a0 in ?? ()
#3  0x0000000000000004 in ?? ()
#4  0x0000000000000090 in ?? ()
#5  0x0000000001259390 in ?? ()
#6  0x0000000001259410 in ?? ()
#7  0x00000000012593a0 in ?? ()
#8  0x0000000000000080 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb) c
Continuing.
^C
Program received signal SIGINT, Interrupt.
0x000000000022af10 in ?? ()
(gdb) bt
#0  0x000000000022af10 in ?? ()
#1  0x000000000125fd30 in ?? ()
#2  0x00000000000000a0 in ?? ()
#3  0x0000000000000004 in ?? ()
#4  0x0000000000000090 in ?? ()
#5  0x0000000001259390 in ?? ()
#6  0x0000000001259410 in ?? ()
#7  0x00000000012593a0 in ?? ()
#8  0x0000000000000080 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb) c
Continuing.
^C
Program received signal SIGINT, Interrupt.
0x000000000022ae80 in ?? ()
(gdb) bt
#0  0x000000000022ae80 in ?? ()
#1  0x000000000125fd30 in ?? ()
#2  0x00000000000000a0 in ?? ()
#3  0x0000000000000004 in ?? ()
#4  0x0000000000000090 in ?? ()
#5  0x0000000001259390 in ?? ()
#6  0x0000000001259410 in ?? ()
#7  0x00000000012593a0 in ?? ()
#8  0x0000000000000080 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb)

...

$ sha256sum /usr/sbin/init
72190cb5c04e9cc370afa91721d67e070677bbefe874b69dd76ccb86e8188c8c  /usr/sbin/init
```

**Comment by Radrik5** [comment]
I experience slowness in Git Extensions running on Windows with repo in WSL 2, init process in WSL consumes more that 100% CPU.

I found an easy way to reproduce the issue: clone WSL2-Linux-Kernel and run 'git status' from Windows.

In WSL2 Ubuntu:
```
$ git clone -b linux-msft-wsl-5.10.y https://github.com/microsoft/WSL2-Linux-Kernel
```

In Windows console:
```
> subst x: \\wsl$\Ubuntu
> x:
> cd home\<user>\WSL2-Linux-Kernel
> git status
```

![image](https://user-images.githubusercontent.com/9339459/130315897-cd5d4af8-3fc4-4298-8521-a55afa0d6763.png)

**Comment by derekrprice** [comment]
@tuhlmann said:
> I experience the same thing on my wsl2 Ubuntu 18.04 distro. It _seems_ to be related to a process from Windows, like IntelliJ Idea, to scan through its project files and re-index them, at least these are the occasions at which I observed this behavior. Shutting down Idea doesn't help.

I am seeing that with Ubuntu 20,04.  I closed PHPStorm (closely related to IntelliJ), and the problem went away.  I think this is their problem.

**Comment by lukasamd** [comment]
@derekrprice yep, it's their problem.. or just problem with communication between windows and WSL filesystem
I've used phpStorm from WSL instead of Windows for about 2 months and didn't see that issue anymore

**Comment by PioPh** [comment]
I've been having this issue as well.
In my case it seems to be an issue particularly with Windows Terminal (wt.exe) opening wsl as the default profile.
Opening wsl with cmd.exe, powershell.exe or wt.exe with anything else as the default profile then running wsl does not have the same init behaviour.

Windows 10 19044.1415 21H2. WSL 2. Kali-linux.

**Comment by romkatv** [comment]
Just as @PioPh described: opening a WSL tab in Windows Terminal causes 100% CPU usage by `/init`, but opening a Command Prompt tab and manually running `wsl.exe` from it works fine. This is with WSL1. Reboot doesn't help. Reproducible every time. Started happening very recently -- some time this week.

**Comment by marchelzo** [comment]
Yep exactly the same for me. Started about two weeks ago out of nowhere after a Windows update I believe.

**Comment by g4b0** [comment]
Same problem on my workstation, with ubuntu 18 and wsl2. Like @romkatv it just happens opening a tab in Windows Terminal

**Comment by romkatv** [comment]
I've asked my friends and it appears that the problem reproduces for everyone:

1. Make Ubuntu 20.04 the default distro in WSL (either WSL1 or WSL2 -- doesn't matter).
2. Open a new WSL tab in Windows Terminal.
3. Observe that one CPU core is now at 100% load.

The version of Ubuntu doesn't seem to matter. It's possible that the choice of distro also doesn't matter.

**Comment by myndbreaker** [comment]
I had success fixing this with resetting the Userhome Path in Windows terminal to default and then back to my wsl Userhome.
The only Difference I could spot with the userhome Path was the Change from Slashes to Backslashes 
\\wsl$\Debian\home\USERNAME

**Comment by romkatv** [comment]
Confirmed. Opening a WSL tab in Windows Terminal when `startingDirectory` is set to `/home/romka` results in 100% load of one CPU core. Doing the same when `startingDirectory` is `\\wsl$\Ubuntu-20.04\home\romka` works fine.

I've changed `startingDirectory` from `//wsl$/Ubuntu-20.04/home/romka` to `/home/romka` recently, when the former stopped working. I suppose that's when 100% CPU load started happening.

**Comment by colomb8** [comment]
quick and dirty fix, waiting for the real fix:

- comment "startingDirectory" in settings.json
- add "cd ~" at the end of .bashrc

**Comment by cmsavage** [comment]
Confirming that the Windows Terminal starting directory was the issue for me (WSL, Ubuntu 18.04): changing from `//wsl$/Ubuntu-18.04/home/username` to `\\wsl$\Ubuntu-18.04\home\username` stopped the excess CPU usage.

If you use the "open windows from previous session" as your Windows Terminal startup mode, the issue will come back for any sessions that you started prior to fixing the starting directory.  Close each of those and start new sessions once, then they should work correctly upon Windows Terminal restart.

**Comment by sorgloomer** [comment]
Fyi on none of my machines was I able to reproduce the issue for the past couple of months. High CPU usage stops after 1-2 minutes, it is practically indistinguishable from boot load in my case now (previously it didn't stop on itself, not even after 30 minutes).

**Comment by kstauffer** [comment]
I think I've figured out the problem. I don't know what's wrong with Windows Terminal (I'm using Preview 1.12.3472.0), but here's how to work around it.

### TL;DR

In your profile, the starting directory must `null` a UNC path -- no forward slashes allowed! `null` will put you in /mnt/c/WINDOWS/System32, which is weird. I suggest you use your WSL root, like `\\wsl$\Ubuntu-20.04`. This is the path used when you start Terminal, or open a new window with `ctrl+shift+n`, or open a new tab with `ctrl+shift+t`. (If you use the `duplicateTab` action instead, make sure you read the next section). 

### Fix your shell

Now, to fix how the shell is reporting the current working directory to Terminal. This is the directory that is used by Terminal for the `duplicateTab` action, overriding the value of `startingDirectory`. I used to have this in my .bashrc: `PROMPT_COMMAND="echo -n \$'\e]9;9;'\"\$PWD\"\$'\e\\\\'"`
Of course, $PWD is a WSL path. The fix is to map that WSL path to its UNC equivalent, with backslashes (not forward!): replace `$PWD` with `wslpath -w $PWD`. Suitably escaped for the various levels of quoting and evaluation, you get: `PROMPT_COMMAND="echo -n \$'\e]9;9;'\"\$(wslpath -w \$PWD)\"\$'\e\\\\'"`

Make sure you don't also have a `\e]9;9;...` sequence in your PS1, or adjust it accordingly.

### Details

If the `startingDirectory` in your profile (or the path most recently reported by the shell via `\e]9;9;` at the moment you invoke `duplicateTab`) begins with a forward /, then you are in for a rough ride.  Terminal will interpret a leading / as a WSL path.
* If that path is "/path/to/exists", your shell will start in that directory, but init will spin at 100%.
* If that path is "/", Terminal will actually use C:\ (at least for me), which inside WSL is /mnt/c. CPU will be at 0%. (Yay? But you're not in /, you're in /mnt/c, so that's weird)
* If that path is "/path/to/bad" which doesn't exist in WSL, your shell will start in "/" instead (actually / this time, not /mnt/c) and init spins at 100%. This also covers the case where people were using bogus "UNC" paths with forward slashes, like `//wsl$/Ubuntu-20.04`.

**Comment by kstauffer** [comment]
The handling of the 9;9; escape is fixed in Terminal Preview 1.13. You can use WSL path directly, so you don't need to use `wslpath -w $PWD` to map it to a UNC directory. The rules (and failure modes) about setting `startingDirectory` in your profile still apply, though.

**Comment by daiwei233** [comment]
I hava the same problem, and I didn't use any software about `Intellij`.
I found some other ways, and it works fine for me:

1.Open the terminal with ubuntu 20.04
2.`ssh {any host}`
3.Kill the init process by  `windows Task Manager`

Now the problem is I can't use `ctrl+c`.

**Comment by sorgloomer** [comment]
Something must have happened, the 100% cpu init process came back and I cannot even kill it anymore

**Comment by vvaltchev** [comment]
I agree with @kstauffer that the problem is related with the starting path. Init began to consume 100% of one CPU after I changed the starting path to /home/vlad. Therefore, I restored the old value and I just added a simple "cd" at the end of my .bashrc file. That was my workaround.

**Comment by kstauffer** [comment]
@vvaltchev Did you try setting the starting path to the Windows equivalent of `\\wsl$\Ubuntu-20.04\home\vlad` (backslashes and the odd prefix included)? Change Ubuntu-20.04 to whatever distro you're using. That fixed it for me, it would be interesting to see if that also works for you.

**Comment by vvaltchev** [comment]
@kstauffer I just did. Using `\\wsl$\Ubuntu-20.04\home\vlad` works well, as expected. `Init` doesn't burn the CPU. Just, it's syntax pretty hard to remember 😄

**Comment by Tofandel** [comment]
It seems there is two problems then, one with `intellij` and one with `startDirectory`, I'm experiencing the first but not the later

So if you have a problem with high CPU and are running an instance of PhpStorm WebStorm or any other intellij product from time to time, some git tasks or indexing task seems to get stuck, in which case restart the stuck instance is enough to solve it

Can't speak for `startDirectory`

**Comment by Robula** [comment]
> It seems there is two problems then, one with `jetbrains` and one with `startDirectory`, I'm experiencing the first but not the later
> 
> So if you have a problem with high CPU and are running an instance of PhpStorm WebStorm or any other jetbrains product from time to time, some git tasks or indexing task seems to get stuck, in which case restart the stuck instance is enough to solve it
> 
> Can't speak for `startDirectory`

I use JetBrains products also and I'm having the same issue with `/init` running away on a single core, I also notice Git inside the JetBrains IDEs either crashes or is just incredibly slow. Have these issues been reported on https://youtrack.jetbrains.com/issues?

**Comment by derekrprice** [comment]
@Robula, I found this one: https://youtrack.jetbrains.com/issue/CPP-23933/WSL-1-init-process-consumes-a-lot-of-CPU

**Comment by Tofandel** [comment]
Hit this again today and no luck killing it this time, I had to restart the LXSSManager service

What's even more scary is that it doesn't show up in the Task manager and this cpu usage is only visible with htop in wsl, so my computer was getting crazy hot to the point of burning and I couldn't figure out why

**Comment by Tofandel** [comment]
@sorgloomer it's unlikely this will solve the issue, as like you I'm running jetbrains from Windows on a project in WSL, but I've already always been using the git from WSL and not Windows and still had this issue

**Comment by TonalidadeHidrica** [comment]
I ran into the same issue (busy `init` process of pid 1) but neither JetBrains nor `startingdirectory` seems the root issue†.  I ran
```
taskkill /IM explorer.exe /F
explorer.exe
```
successivel then the rogue `init` process immediately stopped consuming 44% CPU.  (Yes, I delibarately used a codeblock to catch the eyes who are seeking for another solution ;)  I came up with an idea of killing explorer because, on one hand you guys mentioned that it is due to disk access, and on the other hand I've tried copying a relatively large directory from WSL to C drive, which ended up in a stop of copying halfway done, before which I responded to a strange warning of duplicated files (it shuold not happen when copying a directoly alone, right?) with admitting overwriting files.

† I am not using JetBrains at all, and also I doubt it is due to Windows Terminal simply because such an issue never happened during my few months experience in Windows terminal.

**Comment by jarppiko** [comment]
I came by the same issue. In my case the setup was the following:

```
> wsl --version
WSL version: 1.0.3.0
Kernel version: 5.15.79.1
WSLg version: 1.0.47
MSRDC version: 1.2.3575
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.963
```

Linux OS was Ubuntu 22..04.1 updated from 20.04 through `do-release-upgrade`. In my case the `init` process got triggered by `systemd = true` setting in `/etc/wsl.conf`. 

I managed to overcome the issue by making a fresh install of Ubuntu 22.04.1 LTS from Microsoft Store as a second instance. That does not have any issues with the `systemd = true` setting. I am now in process of migrating my VM content to the new instance.
